src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -3,12 +3,47 @@
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     internal static class CodeRefactoringHelpers
     {
+
+        /// <summary>
+        /// <para>
+        /// Returns an instance of <typeparamref name="TSyntaxNode"/> for refactoring given specified selection in document or null
+        /// if no such instance exists.
+        /// </para>
+        /// <para>
+        /// A <typeparamref name="TSyntaxNode"/> instance is returned if selection is zero-width and inside/touching a Token whose
+        /// direct parent is of type <typeparamref name="TSyntaxNode"/> or if a Node of said type is the smallest Node containing
+        /// the whole <paramref name="selection"/>. Otherwise returns <code>null</code>.
+        /// </para>
+        /// <para>
+        /// Note: this function strips all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The stripped version is then used to determine relevant <see cref="SyntaxNode"/>.
+        /// </para>
+        /// </summary>
+        public static async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken, bool traverseUpInSyntaxTreeIfOnEdgeOfSpan = false) where TSyntaxNode : SyntaxNode
```

Comment:
i'd prefer the parameter be non-optiona and before the cancellation token.  it would be good for people to have to think if they want this (and don't all callers want to support 'true'?).

Reply:
Will fix, good points. 

PS: The parameter actually wasn't supposed to get commited yet (haven't finished the work on it before I got home).

Reply:
That was actually the idea why it was optional, to make guidance for what is expected but still allow a way to "override it".

Reply:
First: let's only add the parameter later if necessary.  
Second, if we want a non-last-parameter to be optional, an easy way to do that is through:

```c#
void Foo(int x, CancellationToken ct) => Foo(x, y: false, ct);
void Foo(int x, int y, CancellationToken ct) => ...
```

This is cleaner and keeps the CT non-optional and always at the end.